### PR TITLE
Improve guide for development with CLion

### PR DESCRIPTION
**Pull request by badlogic**
This improved guide describes how to use the compilation database generated by SCons to work on Godot with CLion. It also describes a setup that allows single click cleaning/building/debugging/profiling of the Godot editor directly from within CLion. As a heavy CLion user, I can confirm that this should be the preferred workflow when using CLion for Godot development.

### CODE REVIEW

File: development/cpp/configuring_an_ide/clion.rst [hunk 1]
```diff
@@ -3,69 +3,80 @@
...
 
 Importing the project
----------------------
+---------------------------------------------------
 
-CLion requires a ``CMakeLists.txt`` file as a project file, which is problematic
-for Godot because it uses the SCons buildsystem instead of CMake. However,
-there is a ``CMakeLists.txt`` configuration for :ref:`Android Studio <doc_configuring_an_ide_android_studio>`
-which can also be used by CLion.
+CLion can import projects `compilation database file <https://clang.llvm.org/docs/JSONCompilationDatabase.html>`, commonly named `compile_commands.json`. To generate the compilation database file, open the terminal, change into the Godot root directory, and execute:
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:12]:
```suggestion
CLion can import a project's `compilation database file <https://clang.llvm.org/docs/JSONCompilationDatabase.html>`, commonly named `compile_commands.json`. To generate the compilation database file, open the terminal, change to the Godot root directory, and run:
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 2]
```diff
@@ -3,69 +3,80 @@
 CLion
 =====
 
-`CLion <https://www.jetbrains.com/clion/>`_ is a commercial
+`CLion <https://www.jetbrains.com/clion/>`_ is a commercial 
 `JetBrains <https://www.jetbrains.com/>`_ IDE for C++.
 
 Importing the project
----------------------
+---------------------------------------------------
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:10]: reStructuredText **requires** titles underlines to be the same length as the titles themselves:

```suggestion
---------------------
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 3]
```diff
@@ -3,69 +3,80 @@
 CLion
 =====
 
-`CLion <https://www.jetbrains.com/clion/>`_ is a commercial
+`CLion <https://www.jetbrains.com/clion/>`_ is a commercial
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:6]:
```suggestion
`CLion <https://www.jetbrains.com/clion/>`_ is a commercial
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 4]
```diff
@@ -3,69 +3,80 @@
...
-  from the top menu.
-- Navigate to ``<Godot root directory>/platform/android/java/nativeSrcsConfigs`` (the
-  ``CMakeLists.txt`` file is located there) and select it (but *not* the
-  ``CMakeLists.txt`` file itself), then click **OK**.
+::
 
-.. figure:: img/clion_1_open.png
+    $ scons compiledb
+
+Then, open the Godot root directory with CLion. CLion will import the compilation database, index the codebase, and provide auto-completion and other advanced code navigation and refactoring functionality.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:18]:
```suggestion
Then, open the Godot root directory with CLion. CLion will import the compilation database, index the codebase, and provide autocompletion and other advanced code navigation and refactoring functionality.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 5]
```diff
@@ -3,69 +3,80 @@
...
+-----------------------------------
+
+CLion does not support compiling and debugging Godot via SCons out of the box. This can be achived by creating a custom build target and run configuration in CLion. Before creating a custom build target, you must `compile Godot <https://docs.godotengine.org/en/stable/development/compiling/index.html>`__ once on the command line, to generate the Godot executable. Open the terminal, change into the Godot root directory, and execute:
+
+::
+
+    $ scons target=debug
+
+
+To add a customer build target that invokes SCons for compilation:
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:30]:
```suggestion
To add a custom build target that invokes SCons for compilation:
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 6]
```diff
@@ -3,69 +3,80 @@
...
+
+- Open CLion, and navigate to ``Preferences > Build, Execution, Deployment > Custom Build Targets``
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:32]:
```suggestion
- Open CLion and navigate to **Preferences > Build, Execution, Deployment > Custom Build Targets**
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 7]
```diff
@@ -3,69 +3,80 @@
...
+
+.. figure:: img/clion-preferences.png
    :align: center
 
-   The folder containing the ``CMakeLists.txt`` file.
+- Click ``Add target`` and give the target a name, e.g. ``Godot debug``.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:37]:
```suggestion
- Click **Add target** and give the target a name, e.g. ``Godot debug``.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 8]
```diff
@@ -3,69 +3,80 @@
...
+
+.. figure:: img/clion-target.png
+   :align: center
 
-- If this popup window appears, select **This Window** to open the project:
+- Click ``...`` next to the `Build:` selectbox, then click the ``+`` button in the ``External Tools`` dialog to add a new external tool.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:42]:
```suggestion
- Click **...** next to the **Build:** selectbox, then click the **+** button in the **External Tools** dialog to add a new external tool.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 9]
```diff
@@ -3,69 +3,80 @@
...
 
-.. figure:: img/clion_2_this_window.png
+.. figure:: img/clion-external-tools.png
    :align: center
 
-- Choose **Tools > CMake > Change Project Root** from the top menu and select
-  the Godot root folder.
+- Give the tool a name, e.g. ``Build Godot debug``, set ``Program`` to ``scons``, set ``Arguments`` to the compilation settings you want (see `Compiling Godot <https://docs.godotengine.org/en/stable/development/compiling/index.html>`__), and set the ``Working directory`` to ``$ProjectFileDir$``, which equals the Godot root directory. Click ``OK`` to create the tool.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:47]:
```suggestion
- Give the tool a name, e.g. ``Build Godot debug``, set **Program** to ``scons``, set **Arguments** to the compilation settings you want (see :ref:`toc-devel-compiling`), and set the **Working directory** to ``$ProjectFileDir$``, which equals the Godot root directory. Click **OK** to create the tool.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 10]
```diff
@@ -3,69 +3,80 @@
...
 
-.. figure:: img/clion_3_change_project_root.png
+.. figure:: img/clion-create-build-tool.png
    :align: center
 
-- You should be now be able to see all the project files. Autocomplete should
-  work once the project has finished indexing.
+- Back in the ``External Tools`` dialog, click the ``+`` again to add a second external tool for cleaning the Godot build via ``Scons``. Give the tool a name, e.g. ``Clean Godot debug``, set ``Program`` to ``scons``, set ``Arguments`` to ``-c`` (which will clean the build), and set the ``Working directory`` to ``$ProjectFileDir$``. Click ``OK`` to create the tool.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:52]:
```suggestion
- Back in the **External Tools** dialog, click the **+** again to add a second external tool for cleaning the Godot build via SCons. Give the tool a name, e.g. ``Clean Godot debug``, set **Program** to ``scons``, set **Arguments** to ``-c`` (which will clean the build), and set the **Working directory** to ``$ProjectFileDir$``. Click **OK** to create the tool.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 11]
```diff
@@ -3,69 +3,80 @@
...
 
-If you run into any issues, ask for help in one of
-`Godot's community channels <https://godotengine.org/community>`__.
+.. figure:: img/clion-create-clean-tool.png
+   :align: center
 
-Debugging the project
----------------------
+- Close the ``External Tools`` dialog. In the ``Custom Build Target`` dialog for the custom ``Godot debug`` build target, select the ``Build Godot debug`` tool from the ``Build`` select box, and select the ``Clean Godot debug`` tool from the ``Clean`` select box. Click ``OK`` to create the custom build target.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:57]:
```suggestion
- Close the **External Tools** dialog. In the **Custom Build Target** dialog for the custom ``Godot debug`` build target, select the **Build Godot debug** tool from the **Build** select box, and select the **Clean Godot debug** tool from the **Clean** select box. Click **OK** to create the custom build target.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 12]
```diff
@@ -3,69 +3,80 @@
...
 
-Since CLion does not support SCons, you won't be able to compile, launch, and debug Godot from CLion in one step.
-You will first need to :ref:`compile Godot yourself <toc-devel-compiling>` and run the binary without CLion. You will then be able to debug Godot by using the `Attach to process <https://www.jetbrains.com/help/clion/attaching-to-local-process.html>`__ feature.
+.. figure:: img/clion-select-tools.png
+   :align: center
 
-- Run the compilation in debug mode by entering ``scons``.
+- In the main IDE window, click ``Add Configuration``.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:62]:
```suggestion
- In the main IDE window, click **Add Configuration**.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 13]
```diff
@@ -3,69 +3,80 @@
...
 
-- Run the binary you have created (in the bin directory). If you want to debug a specific project, run the binary with the following arguments : ``--editor --path path/to/your/godot/project``. To run the project instead of editing it, remove the ``--editor`` argument.
+.. figure:: img/clion-add-configuration.png
+   :align: center
 
-- In CLion, go to **Run > Attach to Process...**
+- In the ``Run/Debug Configuration`` dialog, click ``Add new...``, then select ``Custom Build Application`` to create a new custom run/debug configuration.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:67]:
```suggestion
- In the **Run/Debug Configuration** dialog, click **Add new...**, then select **Custom Build Application** to create a new custom run/debug configuration.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 14]
```diff
@@ -3,69 +3,80 @@
...
 
-.. figure:: img/clion_4_select_attach_to_process.png
+.. figure:: img/clion-add-custom-build-application.png
    :align: center
 
-- Find and Select godot in the list (or type the binary name/Process ID)
+- Give the run/debug configuration a name, e.g. ``Godot debug``, select the ``Godot debug`` custom build target as the ``Target``, select the Godot executable in the ``bin/`` folder as the ``Executable``, and set the ``Program arguments`` to ``--editor --path path-to-your-project/``, where ``path-to-your-project/`` should be a path pointing to an existing Godot project. If you omit the ``--path`` argument, you will only be able to debug the Godot project manager window. Click ``OK`` to create the run/debug configuration.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:72]:
```suggestion
- Give the run/debug configuration a name, e.g. ``Godot debug``, select the ``Godot debug`` custom build target as the **Target**. Select the Godot executable in the ``bin/`` folder as the **Executable**, and set the **Program arguments** to ``--editor --path path-to-your-project/``, where ``path-to-your-project/`` should be a path pointing to an existing Godot project. If you omit the ``--path`` argument, you will only be able to debug the Godot project manager window. Click **OK** to create the run/debug configuration.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 15]
```diff
@@ -3,69 +3,80 @@
...
-You can now use the debugging tools from CLion.
+You can now build, run, debug, profile, and Valgrind check the Godot editor via the run configuration.
 
-.. note::
+.. figure:: img/clion-build-run.png
+   :align: center
 
-    If you run the binary without any arguments, you will only debug the project manager window.
-    Don't forget to add the ``--path path/to/your/godot/project`` argument to debug a project.
+When playing a scene, the Godot editor will spawn a separate process. You can debug this process in CLion by going to ``Run > Attach to process...``, typing ``godot``, and selecting the Godot process with the highest ``pid``, which will usually be the player window.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:82]:
```suggestion
When playing a scene, the Godot editor will spawn a separate process. You can debug this process in CLion by going to **Run > Attach to process...**, typing ``godot``, and selecting the Godot process with the highest **pid** (process ID), which will usually be the running project.
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 16]
```diff
@@ -9,63 +9,74 @@ CLion
...
-You can now use the debugging tools from CLion.
+You can now build, run, debug, profile, and Valgrind check the Godot editor via the run configuration.
 
-.. note::
+.. figure:: img/clion-build-run.png
+   :align: center
 
-    If you run the binary without any arguments, you will only debug the project manager window.
-    Don't forget to add the ``--path path/to/your/godot/project`` argument to debug a project.
+When playing a scene, the Godot editor will spawn a separate process. You can debug this process in CLion by going to **Run > Attach to process...**, typing ``godot``, and selecting the Godot process with the highest **pid** (process ID), which will usually be the running project.
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:82]:
```suggestion
When playing a scene, the Godot editor will spawn a separate process. You can debug this process in CLion by going to **Run > Attach to process...**, typing ``godot``, and selecting the Godot process with the highest **pid** (process ID), which will usually be the running project.

```

File: development/cpp/configuring_an_ide/clion.rst [hunk 17]
```diff
@@ -9,63 +9,74 @@ CLion
...
-- Navigate to ``<Godot root directory>/platform/android/java/nativeSrcsConfigs`` (the
-  ``CMakeLists.txt`` file is located there) and select it (but *not* the
-  ``CMakeLists.txt`` file itself), then click **OK**.
-
-.. figure:: img/clion_1_open.png
+CLion can import a project's `compilation database file <https://clang.llvm.org/docs/JSONCompilationDatabase.html>`, commonly named `compile_commands.json`. To generate the compilation database file, open the terminal, change to the Godot root directory, and run:
+
+::
+
+    $ scons compiledb
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:16]:
```suggestion
    $ scons compiledb=yes
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 18]
```diff
@@ -9,63 +9,74 @@ CLion
...
+Then, open the Godot root directory with CLion. CLion will import the compilation database, index the codebase, and provide autocompletion and other advanced code navigation and refactoring functionality.
+
+Compiling and debugging the project
+-----------------------------------
+
+CLion does not support compiling and debugging Godot via SCons out of the box. This can be achived by creating a custom build target and run configuration in CLion. Before creating a custom build target, you must `compile Godot <https://docs.godotengine.org/en/stable/development/compiling/index.html>`__ once on the command line, to generate the Godot executable. Open the terminal, change into the Godot root directory, and execute:
+
+::
+
+    $ scons target=debug
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:27]:
```suggestion
    scons
```

**Comment by Calinou** [comment]
One issue that was raised on the [Godot contributors chat](https://chat.godotengine.org/) is that CLion doesn't seem to expand shell command usage such as `scons -j$(nproc)`. This causes an error as `$(nproc)` will be passed as-is to the SCons command, while it should be converted to a numeric value by the shell.

The documentation should probably have a warning about this, telling the user to replace `-j$(nproc)`, -j%NUMBER_OF_PROCESSORS` or `-j$(sysctl -n hw.logicalcpu)` with a fixed amount of CPU cores like `-j8`.

**Comment by badlogic** [comment]
Thank you for the swift review and sorry for the many edits you had to make.

Bill had a very good suggestion over on the SCons mailing list regarding the `-j` issue: https://pairlist4.pair.net/pipermail/scons-users/2022-April/008841.html

I suppose this could make it into Godot's SConstruct file as a default? I'll add a callout regarding `-j` to this PR for the current setup.

**Review by mhilbrunner — APPROVED** [review]
Nice to see you around these parts :) Looks good to me.

I'll wait for your amendment before merging.
Some of the screenshots seem on the larger side, but if you're busy I can run a compressor on them.

**Comment by badlogic** [comment]
@mhilbrunner @Calinou Sorry for this taking so long. I've now amended the PR as discussed and squashed the commits. Thank you for your review.

File: development/cpp/configuring_an_ide/clion.rst [hunk 19]
```diff
@@ -9,63 +9,76 @@ CLion
...
-
-- From the CLion's welcome window choose the option to import an existing
-  project. If you've already opened another project, choose **File > Open**
-  from the top menu.
-- Navigate to ``<Godot root directory>/platform/android/java/nativeSrcsConfigs`` (the
-  ``CMakeLists.txt`` file is located there) and select it (but *not* the
-  ``CMakeLists.txt`` file itself), then click **OK**.
-
-.. figure:: img/clion_1_open.png
+CLion can import a project's `compilation database file <https://clang.llvm.org/docs/JSONCompilationDatabase.html>`, commonly named `compile_commands.json`. To generate the compilation database file, open the terminal, change to the Godot root directory, and run:
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:12]:
```suggestion
CLion can import a project's `compilation database file <https://clang.llvm.org/docs/JSONCompilationDatabase.html>`, commonly named `compile_commands.json`__. To generate the compilation database file, open the terminal, change to the Godot root directory, and run:
```

File: development/cpp/configuring_an_ide/clion.rst [hunk 20]
```diff
@@ -9,63 +9,76 @@ CLion
...
+-----------------------------------
+
+CLion does not support compiling and debugging Godot via SCons out of the box. This can be achived by creating a custom build target and run configuration in CLion. Before creating a custom build target, you must `compile Godot <https://docs.godotengine.org/en/stable/development/compiling/index.html>`__ once on the command line, to generate the Godot executable. Open the terminal, change into the Godot root directory, and execute:
+
+::
+
+    scons
+
+
+To add a custom build target that invokes SCons for compilation:
```
Comment (Calinou) [inline development/cpp/configuring_an_ide/clion.rst:30]:
```suggestion
::

    scons

To add a custom build target that invokes SCons for compilation:
```

**Review by Calinou — APPROVED** [review]
Thanks! Congratulations for your first merged pull request :tada:

**Pull request merged**